include/beast/core/impl/handler_ptr.ipp

```diff
@@ -0,0 +1,130 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_IMPL_HANDLER_PTR_HPP
+#define BEAST_IMPL_HANDLER_PTR_HPP
+
+#include <boost/asio/detail/handler_alloc_helpers.hpp>
+#include <boost/assert.hpp>
+#include <memory>
+
+namespace beast {
+
+template<class T>
+template<class DeducedHandler, class... Args>
+inline
+handler_ptr<T>::P::
+P(DeducedHandler&& h, Args&&... args)
+    : n(1)
+    , handler(h) // make a copy
+{
+    t = reinterpret_cast<T*>(
+        boost_asio_handler_alloc_helpers::
+            allocate(sizeof(T), h));
+    try
+    {
+        t = new(t) T{h,
+            std::forward<Args>(args)...};
+    }
+    catch(...)
+    {
+        boost_asio_handler_alloc_helpers::
+            deallocate(t, sizeof(T), h);
+        throw;
+    }
+}
+
+template<class T>
+template<class DeducedHandler, class... Args>
+handler_ptr<T>::
+handler_ptr(int, DeducedHandler&& handler, Args&&... args)
+    : p_(new P(std::forward<DeducedHandler>(handler),
+        std::forward<Args>(args)...))
+{
+}
+
+template<class T>
+handler_ptr<T>::
+~handler_ptr()
+{
+    if(! p_)
+        return;
+    if(--p_->n)
+        return;
+    if(p_->t)
+    {
+        p_->t->~T();
+        boost_asio_handler_alloc_helpers::
+            deallocate(p_->t, sizeof(T), p_->handler);
+    }
+    delete p_;
+}
+
+template<class T>
+handler_ptr<T>::
+handler_ptr(handler_ptr&& other)
+    : p_(other.p_)
+{
+    if(! p_)
+        return;
+    other.p_ = nullptr;
+}
+
+template<class T>
+handler_ptr<T>::
+handler_ptr(handler_ptr const& other)
+    : p_(other.p_)
+{
+    if(! p_)
+        return;
+    ++p_->n;
+}
+
+template<class T>
+auto
+handler_ptr<T>::
+release_handler() ->
+    handler_type
+{
+    BOOST_ASSERT(p_);
+    if(p_->t)
+    {
+        p_->t->~T();
+        boost_asio_handler_alloc_helpers::
+            deallocate(p_->t, sizeof(T), p_->handler);
+        p_->t = nullptr;
+    }
+    return std::move(p_->handler);
+}
+
+template<class T>
+template<class... Args>
+void
+handler_ptr<T>::
+invoke(Args&&... args)
+{
+    BOOST_ASSERT(p_);
+    BOOST_ASSERT(p_->t);
+    p_->t->~T();
+    boost_asio_handler_alloc_helpers::
+        deallocate(p_->t, sizeof(T), p_->handler);
+    p_->t = nullptr;
+    p_->handler(std::forward<Args>(args)...);
+}
```

Comment:
Consider implementing this function as `release_handler()(std::forward<Args>(args)...)`. The only difference between the two expressions that I can think of is that `invoke` leaves the handler in place, while `release_handler()(...)` would zap the handler by moving it into its return value, but since both functions first destroy/deallocate `data`, I don't think that'd be a problem. I suppose the difference between the two could be removed by making `release_handler` return an rvalue reference, allowing the recipient of the ref to make the decision to zap or not, but I'm not sure how palatable/correct that is.